cpp/tests/groupby/group_product_test.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <tests/groupby/groupby_test_util.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <cudf/detail/aggregation/aggregation.hpp>
+
+namespace cudf {
+namespace test {
+template <typename V>
+struct groupby_product_test : public cudf::test::BaseFixture {
+};
+
+// These tests will not work for all types until the following ptxas bug is fixed in 10.2
+// https://nvbugswb.nvidia.com/NvBugs5/SWBug.aspx?bugid=3186317&cp=
+// using supported_types = cudf::test::Types<int8_t, int16_t, int32_t, int64_t, float, double>;
+using supported_types = cudf::test::Types<float, double>;
+
+TYPED_TEST_CASE(groupby_product_test, supported_types);
+
+// clang-format off
+TYPED_TEST(groupby_product_test, basic)
+{
+    using K = int32_t;
+    using V = TypeParam;
+    using R = cudf::detail::target_type_t<V, aggregation::PRODUCT>;
+
```

Comment:
You could put these in `groupby_product_test` since you're using the same ones everywhere. You could always subclass `groupby_product_test` to create separate fixtures if you need to change them for specific tests later.

Reply:
To access these from fixture, you would still need to use `TestFixture::R`, `TestFixture::V` which is still verbose.

Reply:
Yes that's fair, just a matter of taste in which you think is more brief/more readable. If you really were aiming for brevity you could define a function
```
groupby_product_test::test(fixed_width_column_wrapper<K> keys, fixed_width_column_wrapper<V> vals, fixed_width_column_wrapper<K> expect_keys, fixed_width_column_wrapper<V> expect_vals)
{
    auto agg = cudf::make_product_aggregation();
    test_single_agg(keys, vals, expect_keys, expect_vals, std::move(agg));
}
```
then just call it in all the tests using initializer lists
```
TYPED_TEST(groupby_product_test, empty_cols)
{
    this->test({}, {}, {}, {});
}
```
That would obscure the meaning of the arguments a little bit, but would remove all the duplication of both `using-declarations` and defining local variables for all those objects in each test. Up to you, choose whichever you find most readable. These are just minor suggestions for simplification that don't affect correctness/performance anyway.

Reply:
keys and vals may not be always fixed_width_column_wrapper.